### PR TITLE
Feat: Improve Dropdown Component

### PR DESCRIPTION
**Pull request by ArthurTriis1**
## What's the purpose of this pull request?

Improve the Dropdown by adding new options and expanding its usability. In this PR, properties like `asChild`, `alignment`, and `dismissOnClick` have been added. There are no breaking changes, but properties related to icons in `DropdownButton` and `DropdownMenuItem` were marked as deprecated, as these functionalities can now be configured directly within the component content.

### New Hooks
As part of the refactor of the `DropdownButton` and `DropdownMenuItem` components, new hooks were created that encapsulate logic, enabling new components to be created with a fresh structure while maintaining functionality. The new hooks are `useDropdownItem` and `useDropdownTrigger`.

### Alignment Property
A new property called `align` has been added to the `DropdownMenu`. This property allows the `DropdownMenu` to align with the `DropdownButton` based on the desired alignment: right, left, or center.

| right     | center | left     |
|----------|--------|------------|
| Screenshot 2024-10-07 at 23 24 07     | Screenshot 2024-10-07 at 23 23 31    | Screenshot 2024-10-07 at 23 24 35     |

### Adjustments to the `useDropdownPosition` Hook
Some adjustments have been made to the `useDropdownPosition` hook. It now observes changes in window resizing, fixing the alignment bug for the Dropdown.

### Controlled Component
New functionality has been added to allow the Dropdown to be controlled externally, enabling it to be manipulated through external states. Example:

```jsx
      const [isOpen, setIsOpen] = useState(false)
      return (
        <Dropdown isOpen={isOpen} onDismiss={() => {setIsOpen(false)}}>
          <DropdownButton>Controlled Dropdown</DropdownButton>
          <DropdownMenu>
            <DropdownItem>A - Dropdown item as child</DropdownItem>
            <DropdownItem>B - Dropdown item as child</DropdownItem>
            <DropdownItem>C - Dropdown item as child</DropdownItem>
            <DropdownItem>D - Dropdown item as child</DropdownItem>
          </DropdownMenu>
        </Dropdown>
```

The usage of the controlled Dropdown can be seen in the preview at: [#controlled](https://faststore-site-git-feat-improve-dropdowneat-faststore.vercel.app/components/molecules/dropdown#controlled)

### asChild Property
Based on libraries like Radix, the DropdownButton and DropdownMenuItem components now include the asChild property. This allows replacing the default rendered components with custom child elements, while maintaining behavior through prop passing. Example:

```jsx
      <Dropdown>
        <DropdownButton asChild>
          <button>{'Dropdown with trigger as child'}</button>
        </DropdownButton>
        <DropdownMenu style={{ backgroundColor: "white"}}>
          <DropdownItem asChild>
            <button>A - Dropdown item as child</button>
          </DropdownItem>
          <DropdownItem asChild>
            <button>B - Dropdown item as child</button>
          </DropdownItem>
          <DropdownItem asChild>
            <button>C - Dropdown item as child</button>
          </DropdownItem>
          <DropdownItem asChild>
            <button>D - Dropdown item as child</button>
          </DropdownItem>
        </DropdownMenu>
      </Dropdown>
```

The usage of the asChild property can be seen in the preview at: [#as-child](https://faststore-site-git-feat-improve-dropdowneat-faststore.vercel.app/components/molecules/dropdown#as-child)

How to test it?
You can test it by visiting the documentation at: https://faststore-site-git-feat-improve-dropdowneat-faststore.vercel.app/components/molecules/dropdown

References
[Dropdown Menu](https://www.radix-ui.com/primitives/docs/components/dropdown-menu)
[B2BTEAM-1895](https://vtex-dev.atlassian.net/browse/B2BTEAM-1895)

## Checklist

<em>You may erase this after checking them all :wink:</em>

**PR Title and Commit Messages**

- [x] PR title and commit messages follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) specification
  - Available prefixes: feat, fix, chore, docs, style, refactor and test

**PR Description**

- [x] Added a label according to the PR goal - breaking change, bug, contributing, performance, documentation..

**Dependencies**

- [ ] Committed the yarn.lock file when there were changes to the packages

[B2BTEAM-1895]: https://vtex-dev.atlassian.net/browse/B2BTEAM-1895?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/site/middleware.ts [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
+  //     )}`
+  //   )
+  // }
+  // if (pathname.startsWith('/components')) {
+  //   return NextResponse.redirect(
+  //     `https://developers.vtex.com/docs/guides/faststore${pathname
+  //       .replace('/components', '')
+  //       .replace(/\/([^\/]*)\//, '/$1-')}`
+  //   )
+  // }
```
Comment (ArthurTriis1) [inline apps/site/middleware.ts:19]: Uncomment before the merge

File: apps/site/pages/components/molecules/dropdown.mdx [hunk 2]
```diff
@@ -231,6 +232,71 @@ Follow the instructions in the [Importing FastStore UI component styles](/docs/c
...
+        <DropdownItem  asChild>
+          <button>C - Dropdown item as child</button>
+        </DropdownItem>
+        <DropdownItem  asChild>
+          <button>D - Dropdown item as child</button>
+        </DropdownItem>
+      </DropdownMenu>
+    </Dropdown>
+  )
+}
```
Comment (ArthurTriis1) [inline apps/site/pages/components/molecules/dropdown.mdx:260]: Adds controlled option for Dropdown Component

File: packages/components/src/molecules/Dropdown/DropdownMenu.tsx [hunk 3]
```diff
@@ -89,25 +95,56 @@ const DropdownMenu = ({
...
+  
+    const matchItem = rearrangedDropdownItems.find(
+      (item) => item.textContent?.[0]?.toLowerCase() === key.toLowerCase()
+    );
+  
+    if (matchItem) {
+      selectedDropdownItemIndexRef!.current = dropdownItems.indexOf(matchItem);
+      matchItem.focus();
+    }
+  };
```
Comment (ArthurTriis1) [inline packages/components/src/molecules/Dropdown/DropdownMenu.tsx:118]: Add key navigation to Dropdown component

File: packages/components/src/molecules/Dropdown/hooks/useDropdownItem.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+export type UseDropdownItemProps<
+  E extends HTMLElement = HTMLElement,
+> = {
+  ref: React.ForwardedRef<E>
+  onClick?: React.MouseEventHandler<E>
+  dismissOnClick?: boolean
+}
+
+export const useDropdownItem = <
```
Comment (ArthurTriis1) [inline packages/components/src/molecules/Dropdown/hooks/useDropdownItem.ts:13]: Extract the logic of DropdownItem to a reusable hook

File: packages/components/src/molecules/Dropdown/hooks/useDropdownPosition.ts [hunk 5]
```diff
@@ -1,33 +1,60 @@
...
+      const topPosition = topLevel + topOffset + scrollTop
+
+      const leftPosition = leftLevel + scrollLeft
+
+      setPositionProps({
+        top: topPosition,
+        left: leftPosition,
+        loading: false
+      })
+    }
```
Comment (ArthurTriis1) [inline packages/components/src/molecules/Dropdown/hooks/useDropdownPosition.ts:44]: Update Dropdown position based on window resize

File: packages/components/src/molecules/Dropdown/hooks/useDropdownTrigger.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  ])
+
+  return {
+    onClick: toggle,
+    ref: addDropdownTriggerRef,
+    'aria-expanded': isOpen,
+    'aria-controls': id,
+    'aria-haspopup': 'menu' as const,
+  }
+}
```
Comment (ArthurTriis1) [inline packages/components/src/molecules/Dropdown/hooks/useDropdownTrigger.ts:26]: Extract the logic of DropdownButton to a reusable hook

File: packages/components/src/atoms/Button/index.ts [hunk 7]
```diff
@@ -1,2 +1,2 @@
 export { default } from './Button'
-export type { ButtonProps } from './Button'
+export type { ButtonProps, Variant, Size, IconPosition } from './Button'
```
[the remark this answers is not in the document]
Comment (ArthurTriis1) [reply packages/components/src/atoms/Button/index.ts:2]:
```suggestion
export type { ButtonProps, Variant as ButtonVariant, Size as ButtonSize, IconPosition as ButtonIconPosition } from './Button'
```

File: packages/components/src/molecules/Dropdown/DropdownItem.tsx [hunk 8]
```diff
@@ -10,63 +10,47 @@ export interface DropdownItemProps
...
+   * @deprecated
    * A React component that will be rendered as an icon.
    */
   icon?: ReactNode
+  /** 
+   * Replace the default rendered element with the one passed as a child, merging their props and behavior. 
+   * */
+  asChild?: boolean
+  /**
+   *  Emit onDismiss eventwhen the component is clicked.
```
Comment (lariciamota) [inline packages/components/src/molecules/Dropdown/DropdownItem.tsx:22]:
```suggestion
   * Emit onDismiss event when the component is clicked.
```

File: packages/components/src/molecules/Dropdown/DropdownButton.tsx [hunk 9]
```diff
@@ -1,51 +1,64 @@
...
+  /**
+   * @deprecated
+   * A React component that will be rendered as an icon.
+   */
+  icon?: ReactNode
+  /**
+   * @deprecated
+   * Specifies where the icon should be positioned
    */
-  'aria-label'?: AriaAttributes['aria-label']
```
Comment (lariciamota) [inline packages/components/src/molecules/Dropdown/DropdownButton.tsx:15]: question: was it handled somewhere else?

File: packages/components/src/molecules/Dropdown/DropdownButton.tsx [hunk 10]
```diff
@@ -1,51 +1,64 @@
-import React, { forwardRef, useImperativeHandle, AriaAttributes } from 'react'
-import Button, { ButtonProps } from '../../atoms/Button'
-
-import { useDropdown } from './hooks/useDropdown'
+import React, { cloneElement, forwardRef, ReactNode } from 'react'
+import Button, { ButtonProps, IconPosition } from '../../atoms/Button'
```
Comment (lariciamota) [inline packages/components/src/molecules/Dropdown/DropdownButton.tsx:2]:
```suggestion
import Button, { ButtonProps, ButtonIconPosition } from '../../atoms/Button'
```

File: packages/components/src/molecules/Dropdown/DropdownButton.tsx [hunk 11]
```diff
@@ -1,51 +1,64 @@
...
+   * @deprecated
+   * A React component that will be rendered as an icon.
+   */
+  icon?: ReactNode
+  /**
+   * @deprecated
+   * Specifies where the icon should be positioned
    */
-  'aria-label'?: AriaAttributes['aria-label']
+  iconPosition?: IconPosition
```
Comment (lariciamota) [inline packages/components/src/molecules/Dropdown/DropdownButton.tsx:32]:
```suggestion
  iconPosition?: ButtonIconPosition
```

**Review by lariciamota — DISMISSED** [review]

File: packages/components/src/molecules/Dropdown/DropdownButton.tsx (hunk 11 above)
Comment (ArthurTriis1) [reply packages/components/src/molecules/Dropdown/DropdownButton.tsx:32]: Not that i saw, but it still works as expected. The ButtonProps already includes the aria-label prop, so it seems to be redundant."

File: packages/components/src/molecules/Dropdown/DropdownButton.tsx (hunk 9 above)
Comment (ArthurTriis1) [reply packages/components/src/molecules/Dropdown/DropdownButton.tsx:15]: Not that I saw, but it still works as expected. The ButtonProps already includes the aria-label prop, so it seems to be redundant."

**Review by renatomaurovtex — APPROVED** [review]

**Review by renatamottam — APPROVED** [review]

**Pull request merged**